### PR TITLE
settings: API for get storage instance used

### LINKED ISSUE #48328: Add API to get the nvs_fs struct from the settings backend
**Is your enhancement proposal related to a problem? Please describe.**
The problem is that currently it is not possible to use the same storage area for custom NVS usage and the NVS settings backend. The reason for this is because they would need to use the same `nvs_fs struct`, but there are no APIs to fetch this struct from the settings backend.

**Describe the solution you'd like**
I did create an API that did the job, see this Devzone reply: https://devzone.nordicsemi.com/support-private/support/291529#permalink=776251

However, this is probably not an optimal solution. I talked to @nvlsianpu, who suggested an API function that fetches *void pointer to storage which is used by settings, so something like this would be nice.

### PR DESCRIPTION
**Pull request by nvlsianpu**
Added API function +int settings_storage_get(void **storage)
which allows to get storage instance used by the
settings backed to store its records.

fixes #48328

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

**Comment by nvlsianpu** [comment]
cc @simon-iversen

File: include/zephyr/settings/settings.h [hunk 1]
```diff
@@ -585,6 +592,18 @@ int settings_runtime_commit(const char *name);
 
 #endif /* CONFIG_SETTINGS_RUNTIME */
 
+/**
+ * Get the storage instance used by zephyr.
+ *
+ * The type of storage object instance depends on the settings backend used.
+ * It might pointer to: `struct nvs_fs`, `struct fcb` or string witch file name
+ * depens on settings backend type used.
```
Comment (nordicjm) [inline include/zephyr/settings/settings.h:600]: *which
*depends

File: include/zephyr/settings/settings.h [hunk 2]
```diff
@@ -585,6 +592,18 @@ int settings_runtime_commit(const char *name);
...
+ *
+ * @retval Pointer to which reference to the storage objest can be stored.
```
Comment (nordicjm) [inline include/zephyr/settings/settings.h:602]: *Object

File: tests/subsys/settings/functional/fcb/settings_test_fcb.c [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#include <zephyr/fs/fcb.h>
+
+void test_setting_storage_get(void)
+{
+	int rc;
+	void *storage;
+	struct fcb_entry loc;
+
+	rc = settings_storage_get(&storage);
+	zassert_equal(0, rc, "Can't featch storage reference (err=%d)", rc);
```
Comment (nordicjm) [inline tests/subsys/settings/functional/fcb/settings_test_fcb.c:17]: *fetch

File: tests/subsys/settings/functional/file/settings_test_fs.c [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+void test_setting_storage_get(void)
+{
+	int rc;
+	void *storage;
+
+	struct fs_dirent entry;
+
+	rc = settings_storage_get(&storage);
+	zassert_equal(0, rc, "Can't featch storage reference (err=%d)", rc);
```
Comment (nordicjm) [inline tests/subsys/settings/functional/file/settings_test_fs.c:18]: *Fetch

File: tests/subsys/settings/functional/nvs/settings_test_nvs.c [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+void test_setting_storage_get(void)
+{
+	int rc;
+	void *storage;
+	uint16_t data = 0x5a5a;
+	ssize_t nvs_rc;
+
+	rc = settings_storage_get(&storage);
+	zassert_equal(0, rc, "Can't featch storage reference (err=%d)", rc);
```
Comment (nordicjm) [inline tests/subsys/settings/functional/nvs/settings_test_nvs.c:18]: *Fetch

File: subsys/settings/src/settings_store.c [hunk 6]
```diff
@@ -158,6 +158,21 @@ int settings_save(void)
 	return rc;
 }
 
+int settings_storage_get(void **storage)
+{
+	struct settings_store *cs = settings_save_dst;
+
+	if (!cs) {
+		return -4;
```
Comment (nordicjm) [inline subsys/settings/src/settings_store.c:166]: defined errno code for magic number?

Comment (nvlsianpu) [reply subsys/settings/src/settings_store.c:166]: that remains after debugging :D - will remove.

**Review by de-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**